library/lib.sh

```diff
@@ -300,20 +310,56 @@ rolesRunPlaybook() {
     local test_playbook=$2
     local inventory=$3
     local skip_tags=$4
+    local limit=$5
     local LOGFILE="${test_playbook%.*}"-ANSIBLE-"$ANSIBLE_VER"
     local result=FAIL
+    local cmd log_msg
+    cmd="ansible-playbook -i $inventory $skip_tags $limit $tests_path$test_playbook -vv"
+    log_msg="Test $test_playbook with ANSIBLE-$ANSIBLE_VER on ${limit/--limit /}"
     # If LSR_TFT_DEBUG is true, print output to terminal
     if [ "$LSR_TFT_DEBUG" == true ] || [ "$LSR_TFT_DEBUG" == True ]; then
-        rlRun "ANSIBLE_LOG_PATH=$LOGFILE ansible-playbook -i $inventory $skip_tags $tests_path$test_playbook -vv && result=SUCCESS" 0 "Test $test_playbook with ANSIBLE-$ANSIBLE_VER"
+        rlRun "ANSIBLE_LOG_PATH=$LOGFILE $cmd && result=SUCCESS" 0 "$log_msg"
     else
-        rlRun "ansible-playbook -i $inventory $skip_tags $tests_path$test_playbook -vv &> $LOGFILE && result=SUCCESS" 0 "Test $test_playbook with ANSIBLE-$ANSIBLE_VER"
+        rlRun "$cmd &> $LOGFILE && result=SUCCESS" 0 "$log_msg"
     fi
     logfile_name=$LOGFILE-$result.log
     mv "$LOGFILE" "$logfile_name"
     LOGFILE=$logfile_name
     rolesUploadLogs "$LOGFILE"
 }
 
+rolesRunPlaybooksParallel() {
+    # Run playbooks on managed nodes one by one
+    # Supports running against a single node too
+    local tests_path=$1
+    local inventory=$2
+    local skip_tags=$3
+    local test_playbooks=$4
+    local managed_nodes=$5
+    local test_playbooks_arr
+
+    mapfile -t test_playbooks_arr <<< "$test_playbooks"
+    while [[ -n "${test_playbooks_arr[*]}" ]]; do
+        for managed_node in $managed_nodes; do
+            if ! pgrep -af "ansible-playbook" | grep -q "\--limit $managed_node\s"; then
+                test_playbook=${test_playbooks_arr[0]}
+                test_playbooks_arr=("${test_playbooks_arr[@]:1}") # Remove first element from array
```

Comment:
you should use `unset test_playbooks_arr[0]` to remove the first element - see https://www.gnu.org/software/bash/manual/html_node/Arrays.html

Reply:
unset doesn't remove the 0 element, it unsets it to an empty string:
```
[spetrosi@fedora tft-tests]$ test_playbooks_arr=("playbook1" "playbook2" "playbook3")
[spetrosi@fedora tft-tests]$ echo ${test_playbooks_arr[@]}
playbook1 playbook2 playbook3
[spetrosi@fedora tft-tests]$ echo ${test_playbooks_arr[0]}
playbook1
[spetrosi@fedora tft-tests]$ unset test_playbooks_arr[0]
[spetrosi@fedora tft-tests]$ echo ${test_playbooks_arr[0]}
[spetrosi@fedora tft-tests]$ test_playbooks_arr=("${test_playbooks_arr[@]:1}")
[spetrosi@fedora tft-tests]$ echo ${test_playbooks_arr[0]}
playbook2
```

Reply:
did not know that - I thought bash arrays had mutable lengths